### PR TITLE
feat(executor): Add Timm image encoder

### LINKED ISSUE #136: Create TimmImageEncoder
[timm](https://github.com/rwightman/pytorch-image-models) is **the largest** library of image models. Not only that, it also has a unified and very simple to use interface, that stays the same across models. For example,

```python
from timm import create_model

# num_classes=0 will get you last layer features
model = create_model('my_model', num_classes=0, pretrained=True)
```
will work with when replacing `my_model` with `resnet34` or `vit_large_patch16_384` - two completely different model architectures.

This is an extremely low hanging fruit, I am shocked it has not been implemented yet (especially since some other models - like BigTransfer, that we have implemented, are just a small subset of what's available in this library).

### PR DESCRIPTION
**Pull request by vivek2301**
closes #136

### CODE REVIEW

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+from timm import create_model, list_models, list_modules
+from timm.data import resolve_data_config
+from timm.data.transforms_factory import create_transform
+
+
+class TimmImageEncoder(Executor):
+    """
+    :class:`TimmImageEncoder` encodes ``Document`` blobs of type `ndarray` (`float32`) and shape
+    `B x H x W x C` into `ndarray` of `B x D`.
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:22]: Information on the `B` dimension not needed. Our documents are individual images, the user does not need to know about the batch dimension. The executor does the batching internally

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    :param default_batch_size: Defines the batch size for inference on the loaded PyTorch model.
+    :param use_default_preprocessing: Should the input be preprocessed with default configuration.
+    :param args:  Additional positional arguments.
+    :param kwargs: Additional keyword arguments.
+    """
+    def __init__(
+        self,
+        model_name: str = 'resnet18',
+        pooled_features: Optional[bool] = True,
+        device: Optional[str] = None,
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:44]: Default should be `cpu`

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        self.default_batch_size = default_batch_size
+        self.use_default_preprocessing = use_default_preprocessing
+
+        self.default_traversal_path = default_traversal_path
+
+        # axis 0 is the batch
+        self._default_channel_axis = 1
+
+        self.is_model_supported(model_name)
+        if pooled_features:
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:65]: Please assume the user always wants pooled features

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            self._model = create_model(model_name, pretrained=True, num_classes=0)
+        else:
+            self._model = create_model(model_name, pretrained=True, num_classes=0, global_pool='')
+
+        config = resolve_data_config({}, model=self._model)
+        self._preprocess = create_transform(**config)
+        self._preprocess.transforms.insert(0, T.ToPILImage())
+
+    def is_model_supported(self, model_name):
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:74]: Is this needed? Is the error message from `timm` not clear enough?

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        self.default_traversal_path = default_traversal_path
+
+        # axis 0 is the batch
+        self._default_channel_axis = 1
+
+        self.is_model_supported(model_name)
+        if pooled_features:
+            self._model = create_model(model_name, pretrained=True, num_classes=0)
+        else:
+            self._model = create_model(model_name, pretrained=True, num_classes=0, global_pool='')
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:68]: Model should immediately set to `eval`

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+        config = resolve_data_config({}, model=self._model)
+        self._preprocess = create_transform(**config)
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:71]: Will these be evaluation trasforms, or train transforms?

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        self._preprocess.transforms.insert(0, T.ToPILImage())
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:72]: Is this needed?

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+            for document_batch in docs_batch_generator:
+                blob_batch = [d.blob for d in document_batch]
+                if self.use_default_preprocessing:
+                    images = np.stack(self._preprocess_image(blob_batch))
+                else:
+                    images = np.stack(blob_batch)
+                
+                tensor = torch.from_numpy(images).to(self.device)
+                features = self._model(tensor)
+                features = features.detach().numpy()
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:110]: You need to use `.cpu()` before `.numpy()`

**Review by tadejsv — CHANGES_REQUESTED** [review]
Good start, but some more work is needed before we can merge this. Apart form the comments I left below, here is what needs to be done:

- remove `.dockerignore`, `.gitignore`, `LICENSE` and `Dockerfile` - recently we've begun to remove these from our folders.
- Use proper linting. Please check again the `CONTRIBUTOR.md` in this repository, the part about linting has been updated. This PR will not pass liniting tests in CI otherwise

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py (hunk 1 above)
Comment (vivek2301) [reply jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:22]: Hi @tadejsv 
I was following ImageTorchEncoder for things like this. I will change this.

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py (hunk 2 above)
Comment (vivek2301) [reply jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:44]: I do set it to "cpu" if device is None. If I set it to "cpu" here then I will have to make it to run on cpu even when gpu is available.

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py (hunk 3 above)
Comment (vivek2301) [reply jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:65]: Okay, I will remove the pooled_features parameter.

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py (hunk 6 above)
Comment (vivek2301) [reply jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:71]: Default value of parameter is_training=False. So, these will be evaluation transforms.

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py (hunk 2 above)
Comment (tadejsv) [reply jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:44]: No, the user can choose where to run it. I don't like the assumption that if GPU is available, it should run on it. This is the user's decision. Please change the default

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py (hunk 7 above)
Comment (vivek2301) [reply jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:72]: The transform compose created by create_transform starts with resize and crop, which will only function on PILImage. Thus, the input first needs to be converted to PIL Image.

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py (hunk 8 above)
Comment (vivek2301) [reply jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:110]: Yes, thanks. Will correct it.

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py (hunk 4 above)
Comment (vivek2301) [reply jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:74]: I checked it and the message from timm is clear enough. I will remove this check.

**Comment by vivek2301** [comment]
@tadejsv 
If I remove Dockerfile then should I remove "usage via docker image" section from README too?
It was added by default when I created the encoder using jinahub.

**Comment by tadejsv** [comment]
@vivek2301 no, we automatically generate the `Dockerfile` in CI and for use with `jinahub` and `jinahub+docker`

**Comment by vivek2301** [comment]
@tadejsv,

I made the suggested changes and added tests.
Can you please review it?

**Comment by vivek2301** [comment]
@tadejsv 
I've fixed the lint error with isort.
One test failed because the test couldn't pull docker image timmimageencoder. According to contribution guideline, this test needs to be included.
The second test fail is in outer integration test test_dump_psql.py.

Any hints on what should I do to fix these?

**Comment by tadejsv** [comment]
@vivek2301 apologies for not reviewing yet - today is my day off, I will look at it first thing on monday

**Comment by vivek2301** [comment]
No problem, Tadej. Just an update - I noticed that the documentation was updated recently for docker test case and I updated the test accordingly. It should fix the test case.

File: jinahub/encoders/image/TimmImageEncoder/README.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
+# TimmImageEncoder
+
+**TimmImageEncoder** wraps the models from [timm](https://rwightman.github.io/pytorch-image-models/).
+
+**TimmImageEncoder** encodes `Document` blobs of type a `ndarray` and shape Batch x Height x Width x Channel
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/README.md:5]: Remove all mentions of batch dimension

File: jinahub/encoders/image/TimmImageEncoder/README.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+### Inputs 
+If `use_default_preprocessing=True` (recommended):  
+`Document` with `blob` of shape `H x W x C` and dtype `uint8`.  
+
+If `use_default_preprocessing=False`:  
+`Document` with `blob` of shape `C x H x W` and dtype `float32`.
+
+### Returns
+`Document` with `embedding` fields filled with an `ndarray` of the shape `embedding_dim` (size depends on the model) with `dtype=float32`.
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/README.md:89]: Put this on top

File: jinahub/encoders/image/TimmImageEncoder/README.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
+# TimmImageEncoder
+
+**TimmImageEncoder** wraps the models from [timm](https://rwightman.github.io/pytorch-image-models/).
+
+**TimmImageEncoder** encodes `Document` blobs of type a `ndarray` and shape Batch x Height x Width x Channel 
+into a `ndarray` of Batch x Dim and stores them in the `embedding` attribute of the `Document`.
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/README.md:6]: Add initializaion args here (like we have with other executors)

File: jinahub/encoders/image/TimmImageEncoder/manifest.yml [hunk 12]
```diff
@@ -0,0 +1,5 @@
+manifest_version: 1
+name: TimmImageEncoder
+description: '`TimmImageEncoder` encodes data from a `ndarray` into a d-dimensional vector space.'
+url: https://github.com/jina-ai/executors/tree/main/jinahub/encoders/image/TimmImageEncoder
+keywords: [CNN, pytorch, timm, Computer Vision, alexnet, densenet, resnet, efficientnet, mobilenet, ghostnet, image, encode]
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/manifest.yml:5]: remove the model tags

File: jinahub/encoders/image/TimmImageEncoder/manifest.yml [hunk 13]
```diff
@@ -0,0 +1,5 @@
+manifest_version: 1
+name: TimmImageEncoder
+description: '`TimmImageEncoder` encodes data from a `ndarray` into a d-dimensional vector space.'
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/manifest.yml:3]: Can you make this more descriptive of what this executor is?

File: jinahub/encoders/image/TimmImageEncoder/requirements.txt [hunk 14]
```diff
@@ -0,0 +1,7 @@
+jina~=2.0
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/requirements.txt:1]: We've recently started removing jina from requirements

File: jinahub/encoders/image/TimmImageEncoder/requirements.txt [hunk 15]
```diff
@@ -0,0 +1,7 @@
...
+torch==1.9.0
+torchvision==0.10.0
+pillow==8.2
+timm~=0.4.12
+jina-commons @ git+https://github.com/jina-ai/jina-commons.git@v0.0.2#egg=jina-commons
+lz4~=3.1
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/requirements.txt:7]: I don't think this one is needed

File: jinahub/encoders/image/TimmImageEncoder/tests/integration/test_integration.py [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+    with flow:
+        resp = flow.post(
+            on="/test", inputs=[Document(blob=arr_in)], return_results=True
+        )
+
+    results_arr = DocumentArray(resp[0].data.docs)
+    assert len(results_arr) == 1
+    assert results_arr[0].embedding is not None
+    assert results_arr[0].embedding.shape == (512,)
+
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/tests/integration/test_integration.py:29]: No need for this test

```suggestion
```

File: jinahub/encoders/image/TimmImageEncoder/tests/requirements.txt [hunk 17]
```diff
@@ -0,0 +1,4 @@
+pillow==8.2
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/tests/requirements.txt:1]: No need, already in main requirements

File: jinahub/encoders/image/TimmImageEncoder/tests/requirements.txt [hunk 18]
```diff
@@ -0,0 +1,4 @@
...
+pytest==6.2.4
+pytest-mock==3.6.1
+pytest-lazy-fixture==0.6.3
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/tests/requirements.txt:4]: We also don't need them I think?

File: jinahub/encoders/image/TimmImageEncoder/tests/unit/test_unit.py [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+def test_empty_doc_array():
+    docs = DocumentArray()
+    encoder = TimmImageEncoder()
+
+    encoder.encode(docs, parameters={})
+
+    assert len(docs) == 0
+
+
+def test_docs_array_with_no_text():
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/tests/unit/test_unit.py:104]: Docs with no text? Just test with an empty doc. Also, here we are missing a test where `docs=None`

File: jinahub/encoders/image/TimmImageEncoder/tests/unit/test_unit.py [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+
+    for name, image_arr in test_images.items():
+        docs = DocumentArray([Document(blob=image_arr)])
+        encoder.encode(docs, parameters={})
+        embeddings[name] = docs[0].embedding
+
+    def dist(a, b):
+        a_embedding = embeddings[a]
+        b_embedding = embeddings[b]
+        return np.linalg.norm(a_embedding - b_embedding)
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/tests/unit/test_unit.py:68]: Can you rework this test to use `DocumentArray.match` method? See an example in `CLIPImageEncoder`

File: jinahub/encoders/image/TimmImageEncoder/tests/unit/test_unit.py [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+)
+def test_encode_image_returns_correct_length(
+    traversal_paths: Tuple[str], docs: DocumentArray
+) -> None:
+    encoder = TimmImageEncoder(default_traversal_path=traversal_paths)
+
+    encoder.encode(docs=docs, parameters={})
+
+    for doc in docs.traverse_flat(traversal_paths):
+        assert doc.embedding is not None
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/tests/unit/test_unit.py:49]:
```suggestion
```

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+from jina_commons.batching import get_docs_batch_generator
+from timm import create_model
+from timm.data import resolve_data_config
+from timm.data.transforms_factory import create_transform
+
+
+class TimmImageEncoder(Executor):
+    """
+    :class:`TimmImageEncoder` encodes ``Document`` blobs of type `ndarray`
+    (`float32`) and shape `H x W x C` into `ndarray` of `D`.
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:16]: I think you can just put `3` instead of `C`. Also, clarify that the image needs to be in RGB format

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+    If `use_default_preprocessing=False`, the expected input shape is
+    `C x H x W` with `float32` dtype.
+
+    Internally, :class:`TimmImageEncoder` wraps the pre-trained models from
+    `Timm library`.
+    https://rwightman.github.io/pytorch-image-models/
+
+    :param model_name: the name of the model. Models listed on:
+        https://rwightman.github.io/pytorch-image-models/models/
+    :param device: Which device the model runs on. Can be 'cpu' or 'cuda'.
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:28]:
```suggestion
    :param device: Which device the model runs on, for example 'cpu' or 'cuda'.
```

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+    :param default_traversal_paths: Used in the encode method an defines traversal on the received `DocumentArray`.
+    :param default_batch_size: Defines the batch size for inference on the loaded PyTorch model.
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:30]: Can you reword this to make it clearer they are defaults, and can be overwritten?

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+    :param use_default_preprocessing: Should the input be preprocessed with default configuration.
+    :param args:  Additional positional arguments.
+    :param kwargs: Additional keyword arguments.
+    """
+
+    def __init__(
+        self,
+        model_name: str = "resnet18",
+        device: str = "cpu",
+        default_traversal_path: Tuple = ("r",),
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:40]:
```suggestion
        default_traversal_path: Tuple[str] = ("r",),
```

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+    :param default_traversal_paths: Used in the encode method an defines traversal on the received `DocumentArray`.
+    :param default_batch_size: Defines the batch size for inference on the loaded PyTorch model.
+    :param use_default_preprocessing: Should the input be preprocessed with default configuration.
+    :param args:  Additional positional arguments.
+    :param kwargs: Additional keyword arguments.
+    """
+
+    def __init__(
+        self,
+        model_name: str = "resnet18",
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:38]: We use single quotes - can you also change others for this executor? All except for docstrings

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py [hunk 27]
```diff
@@ -0,0 +1,102 @@
...
+        super().__init__(*args, **kwargs)
+
+        self.device = device
+        self.default_batch_size = default_batch_size
+        self.use_default_preprocessing = use_default_preprocessing
+
+        self.default_traversal_path = default_traversal_path
+
+        # axis 0 is the batch
+        self._default_channel_axis = 1
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:55]: Is this used?

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py [hunk 28]
```diff
@@ -0,0 +1,102 @@
...
+                needs_attr="blob",
+            )
+            self._compute_embeddings(docs_batch_generator)
+
+    def _compute_embeddings(self, docs_batch_generator: Iterable) -> None:
+        with torch.no_grad():
+            for document_batch in docs_batch_generator:
+                blob_batch = [d.blob for d in document_batch]
+                if self.use_default_preprocessing:
+                    images = np.stack(self._preprocess_image(blob_batch))
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:90]: No point for creating a special function for this one-liner, let's keep it simple

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py [hunk 29]
```diff
@@ -0,0 +1,102 @@
...
+                else:
+                    images = np.stack(blob_batch)
+
+                tensor = torch.from_numpy(images).to(self.device)
+                features = self._model(tensor)
+                features = features.detach().cpu().numpy()
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:96]:
```suggestion
                features = features.cpu().numpy()
```

**Review by tadejsv — COMMENTED** [review]
Here are some comments. Good job on integrating the docker test!

**Comment by vivek2301** [comment]
Thanks for the review @tadejsv, I will update the PR accordingly tomorrow.

File: jinahub/encoders/image/TimmImageEncoder/tests/requirements.txt (hunk 18 above)
Comment (vivek2301) [reply jinahub/encoders/image/TimmImageEncoder/tests/requirements.txt:4]: pytest-lazy-fixture is used in unit test test_encode_image_returns_correct_length

**Comment by vivek2301** [comment]
@tadejsv 
I made the suggested changes.
Also, I think similar changes are needed for ImageTorchEncoder too. I followed Torch encoder to build Timm encoder and it also has similar issues.

File: jinahub/encoders/image/TimmImageEncoder/manifest.yml [hunk 30]
```diff
@@ -0,0 +1,7 @@
+manifest_version: 1
+name: TimmImageEncoder
+description: '`TimmImageEncoder` encodes data from a `ndarray` into a d-dimensional vector space. It encodes `Document` blobs of shape Height x Width x Channel 
+into a `ndarray` of dimension d and stores them in the `embedding` attribute of the `Document`. The available models is listed in Timm documentation 
+https://rwightman.github.io/pytorch-image-models/ '
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/manifest.yml:5]: Make this description a simple one-liner

File: jinahub/encoders/image/TimmImageEncoder/README.md [hunk 31]
```diff
@@ -0,0 +1,100 @@
...
+from jina import Document, DocumentArray
+from jinahub.image.encoder.timm_encoder import TimmImageEncoder
+
+doc = Document(blob=np.ones((224, 224, 3), dtype=np.uint8))
+encoder = TimmImageEncoder()
+doc_array = DocumentArray([doc, doc])
+encoder.encode(doc_array, parameters={})
+list_embeddings = doc_array.get_attributes('embedding')
+list_embeddings[0].shape, list_embeddings[1].shape
+```
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/README.md:97]: Not needed, you can remove this.
```suggestion
```

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py [hunk 32]
```diff
@@ -0,0 +1,99 @@
...
+        super().__init__(*args, **kwargs)
+
+        self.device = device
+        self.default_batch_size = default_batch_size
+        self.use_default_preprocessing = use_default_preprocessing
+
+        self.default_traversal_path = default_traversal_path
+
+        self._model = create_model(model_name, pretrained=True, num_classes=0)
+        self._model.eval()
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:58]: You need to put the model on `device`

File: jinahub/encoders/image/TimmImageEncoder/tests/unit/test_unit.py [hunk 33]
```diff
@@ -0,0 +1,97 @@
...
+    # without pre-processing the user needs to provide the
+    # right shape for the model directly
+    arr_in = np.ones((3, 224, 224), dtype=np.float32)
+    docs = DocumentArray([Document(blob=arr_in)])
+
+    encoder.encode(docs=docs, parameters={})
+
+    assert docs[0].embedding.shape == (512,)
+
+
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/tests/unit/test_unit.py:85]: Add two tests:
- one where you test different model architectures (3 will be enough)
- one where you test an input of `DocumentArray` with one `Document`, which does not have a `blob` attribute

File: jinahub/encoders/image/TimmImageEncoder/README.md [hunk 34]
```diff
@@ -0,0 +1,100 @@
...
+**TimmImageEncoder** encodes `Document` blobs of type a `ndarray` and shape Height x Width x Channel 
+into a `ndarray` of dimension D and stores them in the `embedding` attribute of the `Document`.
+
+The following parameters can be used:
+
+- `model_name` (string, default: "resnet18"): Name of the pre-trained model
+- `device` (string, default: "cpu"): The device in which the model runs on. For example 'cpu' or 'cuda'
+- `default_traversal_paths` (Tuple(str), default: ("r",)): Traversal path through the docs
+- `default_batch_size` (int, default: 32): Defines the batch size for inference on the loaded Timm model
+- `use_default_preprocessing` (bool, default: True): Should the input be preprocessed with default configuration
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/README.md:14]: You can remove type annotations here

**Review by tadejsv — CHANGES_REQUESTED** [review]
Good effort! Some minor things left for this to be done

**Comment by cristianmtr** [comment]
@vivek2301 @tadejsv what's the status here?

**Comment by tadejsv** [comment]
@vivek2301 seems all my comments were adressed, right?

I see that there is a failure in CI, but that's due to external reasons, I'll investigate

**Comment by vivek2301** [comment]
@tadejsv That's correct, I integrated all your comments.

**Comment by tadejsv** [comment]
@vivek2301, can you merge in the `main` branch? The recent changes on it should allow the CI to pass, so that we can merge this PR

File: jinahub/encoders/image/TimmImageEncoder/__init__.py [hunk 35]
```diff
@@ -0,0 +1 @@
+from .timm_encoder import TimmImageEncoder
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/__init__.py:1]: Just remove this file

File: jinahub/encoders/image/TimmImageEncoder/requirements.txt [hunk 36]
```diff
@@ -0,0 +1,5 @@
+torch==1.9.0
+torchvision==0.10.0
+pillow==8.2
+timm~=0.4.12
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/requirements.txt:4]:
```suggestion
timm==0.4.12
```

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py [hunk 37]
```diff
@@ -0,0 +1,100 @@
...
+        Default batch size is 32 and can be updated by passing an int value.
+    :param use_default_preprocessing: Should the input be preprocessed with default configuration.
+    :param args:  Additional positional arguments.
+    :param kwargs: Additional keyword arguments.
+    """
+
+    def __init__(
+        self,
+        model_name: str = "resnet18",
+        device: str = "cpu",
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:42]:
```suggestion
        device: str = 'cpu',
```

We use single quotes - can you also check other lines in this file.

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py [hunk 38]
```diff
@@ -0,0 +1,100 @@
...
+                needs_attr="blob",
+            )
+            self._compute_embeddings(docs_batch_generator)
+
+    def _compute_embeddings(self, docs_batch_generator: Iterable) -> None:
+        with torch.no_grad():
+            for document_batch in docs_batch_generator:
+                blob_batch = [d.blob for d in document_batch]
+                if self.use_default_preprocessing:
+                    images = np.stack([self._preprocess(img) for img in blob_batch])
```
Comment (tadejsv) [inline jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:91]: Can you reverse the order - first batch, then preprocess. I think torchvision transforms enable that

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py (hunk 37 above)
Comment (tadejsv) [reply jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:42]: If you are using black manually, remember to use `black -S`, so that it won't default to double quotes

**Review by tadejsv — CHANGES_REQUESTED** [review]
Please adress the comments - then I think if the tests pass we can merge

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py (hunk 38 above)
Comment (vivek2301) [reply jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:91]: Stacking it first and then passing the whole batch does not work and ToPILImage transform gives an error as it needs a 2/3 dimensional input. Do you mean stacking it and again looping over the batch to preprocess?

Comment (tadejsv) [reply jinahub/encoders/image/TimmImageEncoder/timm_encoder.py:91]: If it doesn't work, then no problem. Leave it as it is

File: jinahub/encoders/image/TimmImageEncoder/timm_encoder.py (hunk 37 above)
Comment (vivek2301) [reply jinahub/encoders/image/TimmImageEncoder/timm_encoder.py]: Yes, that was the problem. black was replacing the single quotes with double quotes.

**Comment by vivek2301** [comment]
@tadejsv I've merged with the main branch and resolved the comments. Can you please run the workflow?

File: .pre-commit-config.yaml [hunk 39]
```diff
@@ -1,13 +1,28 @@
...
+  hooks:
+  - id: darglint
+    files: jina/
+    exclude: ^(jina/helloworld/|jina/proto/jina_pb2.py|docs/|jina/resources/)
+    args:
+    - --message-template={path}:{line} {msg_id} {msg}
+    - -s=sphinx
+    - -z=full
+    - -v=2
+- repo: https://github.com/pycqa/pydocstyle
```
Comment (tadejsv) [inline .pre-commit-config.yaml:13]: Please revert the changes to this file

**Review by tadejsv — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml (hunk 39 above)
Comment (vivek2301) [reply .pre-commit-config.yaml]: I reverted the changes. I think this file was not there when I forked the repository and I was using another pre-commit config file. I guess it got fetched in this merge and got overwritten. Sorry for that, I should've checked before committing.

**Review by tadejsv — APPROVED** [review]
Good job!
